snapcraft/tests/test_sources.py

```diff
@@ -325,3 +325,16 @@ def test_get_tar_source_from_uri(self):
             with self.subTest(key=source):
                 self.assertEqual(
                     snapcraft.sources._get_source_type_from_uri(source), 'tar')
+
+    def test_get_git_source_from_uri(self):
+        sources = [
+            'git://github.com:ubuntu-core/snapcraft.git',
+            'git@github.com:ubuntu-core/snapcraft.git',
+            'git://github.com/ubuntu-core/snappy-jenkins.git',
+            'git@github.com:ubuntu-core/jenkins-ubuntu.git'
+        ]
+
+        for source in sources:
+            with self.subTest(key=source):
+                self.assertEqual(
+                    snapcraft.sources._get_source_type_from_uri(source), 'git')
```

Comment:
In general we should try to avoid testing internal methods directly, and test that code through the exposed API. In this case, instead of testing  _get_source_type_from_uri, it would be better to test through the get method. You'll have to patch the subprocess.check_call method so Git.pull doesn't contact the URL, and then just check the arguments passed to that check_call. Here are the details about patching: https://docs.python.org/3/library/unittest.mock.html

I like the scenarios and the subTest.